Создание приложения

alex-grt opened the pull request:
- создан файл приложения
- созданы схемы пользователя и карточки фильма
- добавлены роуты
- созданы необходимые контроллеры
- добавлена авторизация
- добавлена валидация
- реализована централизованная обработка ошибок
- подключено логирование
- добавлена обработка CORS

alex-grt committed:
FEATURES

- создан файл приложения
- созданы схемы пользователя и карточки фильма
- добавлены роуты
- созданы необходимые контроллеры
- добавлена авторизация
- добавлена валидация
- реализована централизованная обработка ошибок
- подключено логирование
- добавлена обработка CORS

alex-grt committed:
FIXES

- внесены правки в контроллер удаления карточки фильма

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
    - [x] в проекте есть `package.json`;
    - [x] `.editorconfig`;
    - [x] `.eslintrc`, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
    - [x] `.gitignore`;
    - [x] в `.eslintrc` добавлено исключение для `_id`. Директивы `eslint-disable`, `eslint-disable-line` и `eslint-disable-next-line` запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды `npx eslint .` они должны отсутствовать. 
- [x] <!--3.64--> В разделе scripts файла `package.json` есть команды:
    - [x] `npm run start`, которая запускает сервер на `localhost:3000`;
    - [x] `npm run dev`, которая запускает сервер на `localhost:3000` с хот релоудом.
- [ ] <!--21.81--> Описанные роуты работают корректно: 
    - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
    - [x] `PATCH /users/me` — обновляет информацию о пользователе;
    - [x] `GET /movies` — все сохранённые пользователем фильмы;
    - [ ] `POST /movies` — создаёт фильм с переданными в теле данными;
_- Несоответствие полей в схеме celebrate._
![image](https://user-images.githubusercontent.com/17438227/156421080-26597d3c-1170-46e5-b30f-288d53264cbb.png)
    - [x] `DELETE /movies/:id` — удаляет сохранённый фильм по `_id`;
    - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
    - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
    - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя. 
- [x] <!--3.64--> Все роуты, кроме `/signin` и `/signup`, защищены авторизацией.
- [ ] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
_- signin и signup необходимо описать в отдельном файле._
_- Путь маршрутов необходимо описать в папке routes. В корневом файле нужно подключить маршрут без описания информации о том, при обращении к какому адресу он должен выполняться. На скриншоте указан допустимый вариант подключения маршрута к приложению:_
![image](https://user-images.githubusercontent.com/17438227/132955026-5ef9bc70-c3dd-4811-bef6-7fa3ada24db1.png)
- [ ] <!--7.28--> Ошибки API обрабатываются: 
    - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
    - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
    - [x] сообщение об ошибке соответствует ее типу;
    - [x] асинхронные обработчики завершены блоком `catch`;
    - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
    - [x] пароль хранится в виде хеша;
    - [x] API не возвращает хеш пароля клиенту.
- [ ] <!--7.28--> Валидация реализована верно:
    - [ ] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
_- Запрос на создание пользователя без имени должен прерываться celebrate и не передаваться контроллеру._
    - [ ] Данные валидируются перед добавлением в базу.
_- API позволяет создать пользователя без имени._
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из `process.env`.
- [x] <!--3.64--> К серверу можно обратиться с помощью https по адресу домена, указанному в README.md.
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT: 
    - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
    - [x] в режиме разработки (при  `process.env.NODE_ENV !== 'production'`) код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле `index.js`, который находится в папке `routes`. Оттуда единый роут подключается в файле `app.js`. 
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
    - [x] запросы и ответы записываются в файл `request.log`;
    - [x] ошибки записываются в файл `error.log;`
    - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone.
    Подходящий адрес: movies-explorer.nomoreparties.co/api.
    Неподходящий адрес: movies-explorer.nomoreparties.co.

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор `Error`. 
- [ ] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью. 
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах: 
    - [ ] адрес Mongo-сервера, секретный ключ для JWT в режиме разработки, вынесены в отдельный конфигурационный файл;
    - [ ] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [ ] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено. 
- [ ] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co. 
- [ ] <!--0.71--> Для имени базы данных используется `moviesdb` 

## Количество баллов: 56

## Доп. Комментарии:
- В качестве валидации ссылок в схемах celebrate предлагаю использовать метод `custom` и библиотеку validator https://github.com/hapijs/joi/blob/master/API.md#anycustommethod-description 
![image](https://user-images.githubusercontent.com/17438227/83885018-732f7880-a74e-11ea-905e-28d6e8ef00d1.png)
Это избавит от необходимости писать свои регулярные выражения. 
- при выполнении условия if следующий после условия `if` вызов `next` тоже выполнится. Из-за этого централизованный обработчик ошибок будет пытаться отправить клиенту две ошибки за один запрос.
![image](https://user-images.githubusercontent.com/17438227/120836757-96158800-c56e-11eb-9901-0a870938a524.png)
Ниже пример, который позволит избежать повторной отправки ошибки централизованному обработчику ошибок.
![image](https://user-images.githubusercontent.com/17438227/156421494-1dbe1d35-70aa-40d0-b603-be90d2ef768d.png)
- Ошибка CastError может возникнуть, если пользователь передал неправильный id в запрос. 
Данный запрос не отправляет id от пользователя, то есть данный тип ошибки не сможет возникнуть в запросе. Поэтому проверку на имя `err.name === 'CastError'` можно убрать из условия.
Обработки ошибок 400 следует внедрять в маршруты, которые принимают некоторые данные из body запроса или его параметров. 
![image](https://user-images.githubusercontent.com/17438227/156421673-26a3811d-90f1-4687-86b7-a1e1bdd61204.png)

alex-grt committed:
FIXES

- поле direction исправлено на director в схеме, схеме валидации и контроллере
- маршруты signin и signup вынесены в отдельный файл
- маршруты перенесены из корневого в соответствующие файлы
- поле name в схеме валидации сделано обязательным
- поле name в схеме пользователя сделано обязательным
- в режиме разработки используется devdb, в продакшн - moviesdb (из переменной окружения)
- исправлены условия обработки ошибок, чтобы исключить повторную обработку
- из метода getUser убрана обработка ошибки 400

alex-grt commented:
- поле direction исправлено на director в схеме, схеме валидации и контроллере
- маршруты signin и signup вынесены в отдельный файл
- маршруты перенесены из корневого в соответствующие файлы
- поле name в схеме валидации сделано обязательным
- поле name в схеме пользователя сделано обязательным
- в режиме разработки используется devdb, в продакшн - moviesdb (из переменной окружения)
![db-address](https://user-images.githubusercontent.com/86482022/156599059-1904ac59-ddf8-4f16-b48c-ff4d4ca1d4f9.png)
- исправлены условия обработки ошибок, чтобы исключить повторную обработку
- из метода getUser убрана обработка ошибки 400

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
    - [x] в проекте есть `package.json`;
    - [x] `.editorconfig`;
    - [x] `.eslintrc`, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
    - [x] `.gitignore`;
    - [x] в `.eslintrc` добавлено исключение для `_id`. Директивы `eslint-disable`, `eslint-disable-line` и `eslint-disable-next-line` запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды `npx eslint .` они должны отсутствовать. 
- [x] <!--3.64--> В разделе scripts файла `package.json` есть команды:
    - [x] `npm run start`, которая запускает сервер на `localhost:3000`;
    - [x] `npm run dev`, которая запускает сервер на `localhost:3000` с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно: 
    - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
    - [x] `PATCH /users/me` — обновляет информацию о пользователе;
    - [x] `GET /movies` — все сохранённые пользователем фильмы;
    - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
    - [x] `DELETE /movies/:id` — удаляет сохранённый фильм по `_id`;
    - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
    - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
    - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя. 
- [x] <!--3.64--> Все роуты, кроме `/signin` и `/signup`, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--7.28--> Ошибки API обрабатываются: 
    - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
    - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
    - [x] сообщение об ошибке соответствует ее типу;
    - [x] асинхронные обработчики завершены блоком `catch`;
    - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
    - [x] пароль хранится в виде хеша;
    - [x] API не возвращает хеш пароля клиенту.
- [x] <!--7.28--> Валидация реализована верно:
    - [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
    - [x] Данные валидируются перед добавлением в базу.
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из `process.env`.
- [x] <!--3.64--> К серверу можно обратиться с помощью https по адресу домена, указанному в README.md.
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT: 
    - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
    - [x] в режиме разработки (при  `process.env.NODE_ENV !== 'production'`) код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле `index.js`, который находится в папке `routes`. Оттуда единый роут подключается в файле `app.js`. 
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
    - [x] запросы и ответы записываются в файл `request.log`;
    - [x] ошибки записываются в файл `error.log;`
    - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone.
    Подходящий адрес: movies-explorer.nomoreparties.co/api.
    Неподходящий адрес: movies-explorer.nomoreparties.co.

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор `Error`. 
- [ ] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью. 
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах: 
    - [ ] адрес Mongo-сервера, секретный ключ для JWT в режиме разработки, вынесены в отдельный конфигурационный файл;
    - [ ] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [ ] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено. 
- [ ] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co. 
- [ ] <!--0.71--> Для имени базы данных используется `moviesdb` 

## Количество баллов: 96
